starlette/middleware/limits.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+from typing import ClassVar
+
+import anyio
+
+from starlette.exceptions import HTTPException
+from starlette.responses import PlainTextResponse
+from starlette.types import ASGIApp, Message, Receive, Scope, Send
+
+DEFAULT_MAX_REQUEST_SIZE = 2_621_440  # 2.5MB, same as Django (https://docs.djangoproject.com/en/1.11/ref/settings/#data-upload-max-memory-size)
+
+
+class _TooLarge(HTTPException):
+    msg: ClassVar[str]
+
+    def __init__(self, limit_bytes: int | None) -> None:
+        self.limit = limit_bytes
+        super().__init__(
+            status_code=413,
+            detail=(
+                self.msg + f" Max allowed size is {limit_bytes} bytes."
+                if limit_bytes
+                else self.msg
+            ),
+        )
+
+
+class RequestTooLarge(_TooLarge):
+    """The request body exceeded the configured limit."""
+
+    msg = "Request body is too large."
+
+
+class ChunkTooLarge(_TooLarge):
+    """A chunk exceeded the configured limit."""
+
+    msg = "Chunk size is too large."
+
+
+class RequestSizeLimitMiddleware:
+    def __init__(
+        self,
+        app: ASGIApp,
+        *,
+        max_request_size: int | None = DEFAULT_MAX_REQUEST_SIZE,
+        max_chunk_size: int | None = None,
+        include_limits_in_error_responses: bool = True,
+    ) -> None:
+        self.app = app
+        self.max_request_size = max_request_size
+        self.max_chunk_size = max_chunk_size
+        self.include_limits_in_error_responses = include_limits_in_error_responses
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        total_size = 0
+
+        async def rcv() -> Message:
+            nonlocal total_size
+            message = await receive()
+            chunk_size = len(message.get("body", b""))
+            if self.max_chunk_size is not None and chunk_size > self.max_chunk_size:
+                raise ChunkTooLarge(
+                    self.max_chunk_size
+                    if self.include_limits_in_error_responses
+                    else None
+                )
+            total_size += chunk_size
+            if self.max_request_size is not None and total_size > self.max_request_size:
+                raise RequestTooLarge(
+                    self.max_request_size
+                    if self.include_limits_in_error_responses
+                    else None
+                )
+            return message
```

Comment:
> Then I’m confused as to what you are suggesting: are you saying that we should also enforce that limit, that that limit is already enforced elsewhere, and this PR (or #2174) is not needed, or something else?

Would it be more concise and efficient to change here to judge the request content-length?

Reply:
Instead of the user-defined limits? Or in addition to?

Reply:
What @abersheeran meant is to check the `Content-Length` instead of having the logic of adding up `chunk_size`s - which is actually what Django does: https://github.com/django/django/blob/6daf86058bb6fb922eb3fe3abae6f5c0e645020c/django/http/request.py#L323-L347.

Django also has this logic on the multipart parser: https://github.com/django/django/blob/594873befbbec13a2d9a048a361757dd3cf178da/django/http/multipartparser.py#L241-L248.